broadcaster/backlog/backlog.go

```diff
@@ -0,0 +1,317 @@
+package backlog
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/log"
+	m "github.com/offchainlabs/nitro/broadcaster/message"
+	"github.com/offchainlabs/nitro/util/arbmath"
+)
+
+var (
+	errDropSegments       = errors.New("remove previous segments from backlog")
+	errSequenceNumberSeen = errors.New("sequence number already present in backlog")
+	errOutOfBounds        = errors.New("message not found in backlog")
+)
+
+type Backlog interface {
+	Head() BacklogSegment
+	Append(bm *m.BroadcastMessage) error
+	Get(start, end uint64) (*m.BroadcastMessage, error)
+	Count() int
+}
+
+type backlog struct {
+	head          atomic.Pointer[backlogSegment]
+	tail          atomic.Pointer[backlogSegment]
+	lookupLock    sync.RWMutex
+	lookupByIndex map[uint64]*atomic.Pointer[backlogSegment]
+	config        ConfigFetcher
+	messageCount  atomic.Uint64
+}
+
+func NewBacklog(c ConfigFetcher) Backlog {
+	lookup := make(map[uint64]*atomic.Pointer[backlogSegment])
+	return &backlog{
+		lookupByIndex: lookup,
+		config:        c,
+	}
+}
+
+func (b *backlog) Head() BacklogSegment {
+	return b.head.Load()
+}
+
+// Append will add the given messages to the backlogSegment at head until
+// that segment reaches its limit. If messages remain to be added a new segment
+// will be created.
+func (b *backlog) Append(bm *m.BroadcastMessage) error {
+
+	if bm.ConfirmedSequenceNumberMessage != nil {
+		b.delete(uint64(bm.ConfirmedSequenceNumberMessage.SequenceNumber))
+		// add to metric?
+	}
+
+	// TODO(clamb): Do I need a max catchup config for the backlog? Similar to catchup buffer
```

Comment:
I don't think so, we stopped using that option because we realized it didn't work with client relays